aws/eni.go

```diff
@@ -0,0 +1,206 @@
+package aws
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/microerror"
+)
+
+const (
+	maxRetries    = 15
+	retryInterval = time.Second * 10
+)
+
+type ENIConfig struct {
+	AWSInstanceID string
+	AwsSession    *session.Session
+	DeviceIndex   int64
+	ForceDetach   bool
+	TagKey        string
+	TagValue      string
+}
+
+type ENI struct {
+	awsInstanceID string
+	awsSession    *session.Session
+	deviceIndex   int64
+	forceDetach   bool
+	tagKey        string
+	tagValue      string
+}
+
+func NewENI(config ENIConfig) (*ENI, error) {
+	if config.AWSInstanceID == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.AWSInstanceID must not be empty")
+	}
+	if config.AwsSession == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.AwsSession must not be nil")
+	}
+	if config.DeviceIndex == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "config.DeviceIndex must not be 0")
+	}
+	if config.TagKey == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.TagKey must not be empty")
+	}
+	if config.TagValue == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.TagValue must not be empty")
+	}
+
+	newENI := &ENI{
+		awsInstanceID: config.AWSInstanceID,
+		awsSession:    config.AwsSession,
+		deviceIndex:   config.DeviceIndex,
+		forceDetach:   config.ForceDetach,
+		tagKey:        config.TagKey,
+		tagValue:      config.TagValue,
+	}
+	return newENI, nil
+}
+
+func (s *ENI) AttachByTag() error {
+	// create ec2 client
+	ec2Client := ec2.New(s.awsSession)
+
+	eni, err := s.describe(ec2Client)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	fmt.Printf("Fetched eni-id '%s'\n", *eni.NetworkInterfaceId)
+
+	if *eni.Status == ec2.NetworkInterfaceStatusInUse &&
+		*eni.Attachment.InstanceId == s.awsInstanceID {
+		fmt.Printf("ENI is already attached to this instance. Nothing to do.\n")
+		return nil
+	} else if *eni.Status == ec2.NetworkInterfaceStatusInUse {
+		fmt.Printf("ENI is attached to %q and is in state %q. Trying detach the volume\n", *eni.Attachment.InstanceId, *eni.Status)
```

Comment:
IMO it's not a good practice to print information in random places in a CLI app. I'd rather do a proper error handling and print the the user-facing information in main/cmd package. I understand you put a lot of effort into this already and I think it's quite maintainable so it isn't a blocker for me.

Reply:
thx